includes/class-micropub-media.php

```diff
@@ -0,0 +1,150 @@
+
+		// Pass off to WP to handle the actual upload.
+		$overrides = array(
+			'test_form' => false,
+		);
+
+		// Bypasses is_uploaded_file() when running unit tests.
+		if ( defined( 'DIR_TESTDATA' ) && DIR_TESTDATA ) {
+			$overrides['action'] = 'wp_handle_mock_upload';
+		}
+
+		/** Include admin functions to get access to wp_handle_upload() */
+		require_once ABSPATH . 'wp-admin/includes/admin.php';
+
+		$file = wp_handle_upload( $files[ $name ], $overrides );
+
+		if ( isset( $file['error'] ) ) {
+			return new WP_Micropub_Error( 'invalid_request', $file['error'], 500 );
+		}
+
+		return $file;
+	}
+
+	/**
+	* Check scope
+	*
+	* @param array $scope
+	*
+	* @return boolean
+	**/
+	protected static function check_scopes( $scopes ) {
+		$intersect = array_intersect( $scopes, static::$scopes );
+		return ! empty( $intersect );
+	}
+
+
+	/**
+	 * Checks if a given request has access to create an attachment.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_Micropub_Error|true Boolean true if the attachment may be created, or a WP_Micropub_Error if not.
+	 */
+	protected static function permissions_check( $request ) {
+		if ( ! current_user_can( 'upload_files' ) ) {
+			return new WP_Micropub_Error( 'forbidden', 'User is not permitted to upload files', 403 );
+		}
+		if ( ! self::check_scopes( array( 'create', 'media' ) ) ) {
+			return new WP_Micropub_Error( 'insufficient_scope', 'Token Does Not Meet Requirements for Upload', 401 );
+		}
+
+		return true;
+	}
+
+	public static function upload_handler( $request ) {
+		static::$scopes                 = apply_filters( 'indieauth_scopes', static::$scopes );
+		static::$micropub_auth_response = apply_filters( 'indieauth_response', static::$micropub_auth_response );
+
+		$permission = static::permissions_check( $request );
+		if ( is_micropub_error( $permission ) ) {
+			return $permission;
+		}
+
+		// Get the file via $_FILES
+		$files   = $request->get_file_params();
+		$headers = $request->get_headers();
+		if ( empty( $files ) ) {
+			return new WP_Micropub_Error( 'invalid_request', 'No Files Attached', 400 );
+		} else {
+			$file = self::upload_from_file( $files, 'file' );
+		}
+
+		if ( is_micropub_error( $file ) ) {
+			return $file;
+		}
+
+		$args = array(
+			'post_mime_type' => $file['type'],
+			'guid'           => $file['url'],
+			'post_title'     => preg_replace( '/\.[^.]+$/', '', basename( $file['file'] ) ),
+		);
+
+		$id = wp_insert_attachment( $args, $file['file'], 0, true );
+
+		if ( is_wp_error( $id ) ) {
+			if ( 'db_update_error' === $id->get_error_code() ) {
+				return new WP_Micropub_Error( 'invalid_request', 'Database Error On Upload', 500 );
+			} else {
+				return new WP_Micropub_Error( 'invalid_request', $id->get_error_message(), 400 );
+			}
+		}
+
+		$attachment = get_post( $id );
+
+		// Include admin functions to get access to wp_generate_attachment_metadata().
+		require_once ABSPATH . 'wp-admin/includes/admin.php';
```

Comment:
minor nit: i prefer to do these globally, not inside functions or method. mind moving this to the top?

Reply:
The problem is that globally loads it for the entire WordPress system every time. For some reason, these functions are deliberately not loaded globally at this point in time. They do the exact same thing. I based this function on one built into WordPress itself for their REST API attachment endpoint... https://developer.wordpress.org/reference/classes/wp_rest_attachments_controller/upload_from_file/

Reply:
interesting, ok. out of curiosity, does it cause a problem if you load it globally?

if so, and if this really is the right way, probably add a comment saying what you said here.